### PR TITLE
CI: add TSAN to the sanitizers pipelines

### PR DESCRIPTION
**Pull request by tkf**
This is built on top of #42440 and run `make -C src` with TSAN on CI. It is basically the same as #41530 (plus #41675) but for TSAN instead of ASAN.

Making a draft PR against #42440 so that it'd switch the base to `master` once #42440 is merged.

### CODE REVIEW

**Comment by DilumAluthge** [comment]
I'd like to "promote" asan from the experimental pipeline to the regular pipeline.

Should tsan be experimental, or can it go in the regular pipeline?

**Comment by Keno** [comment]
LGTM

**Comment by tkf** [comment]
I don't have a good sense of how stable TSAN is, since I've never actually made it work locally. That's why I was very eager to add CI, to keep it as close as working possible :stuck_out_tongue:

@Keno @vtjnash @vchuravy I guess you guys know more about if the experimental pipeline is more appropriate for this or not?

**Comment by Keno** [comment]
I don't think we've ever fully been able to run tsan. I think experimental is good for now. Alternatively, we could split it and just have the part that builds the C runtime with tsan in the regular pipeline to ensure that that doesn't regress. Anything beyond that (bootstrapping, maybe running the threads test) can go into the experimental pipeline.

File: contrib/tsan/Make.user.tsan
Comment (Keno) [inline contrib/tsan/Make.user.tsan:10]: This won't work. You need to build LLVM with tsan instrumentation since it uses locks.

Comment (DilumAluthge) [reply contrib/tsan/Make.user.tsan:10]: Yeah I don't want to build LLVM from source in the Base Julia CI. We'd need to have an "LLVM with TSAN" JLL available from Ygg.

Comment (Keno) [reply contrib/tsan/Make.user.tsan:10]: > Can we still use a normal-build LLVM with `deadlock:llvm` in the `TSAN_OPTIONS=suppressions=...` configuration? Or there's no way to work around this fundamentally?

It might be possible, but I'd probably recommend against it. I don't know if that can introduce false positives.

Comment (tkf) [reply contrib/tsan/Make.user.tsan:10]: OK, so maybe it makes sense to close this PR until a JLL is ready.

Or, maybe we can still just run the build process (`-C src`), so that some rudimentary bugs can be caught?

Comment (Keno) [reply contrib/tsan/Make.user.tsan:10]: > Or, maybe we can still just run the build process (`-C src`), so that some rudimentary bugs can be caught?

I think that's a reasonable place to start.

Comment (tkf) [reply contrib/tsan/Make.user.tsan:10]: Actually, the latter part is close to what Keno suggested above https://github.com/JuliaLang/julia/pull/42444#issuecomment-931907446

Comment (DilumAluthge) [reply contrib/tsan/Make.user.tsan:10]: > Or, maybe we can still just run the build process (-C src), so that some rudimentary bugs can be caught?

If this part can be done without needing to build LLVM from source, I think it makes sense to go ahead and add it now.

Comment (DilumAluthge) [reply contrib/tsan/Make.user.tsan:10]: Yeah.

Comment (Keno) [reply contrib/tsan/Make.user.tsan:10]: Yes, no need to instrument LLVM for that.

Comment (DilumAluthge) [reply contrib/tsan/Make.user.tsan:10]: If you can do that in this PR, but leave it as a draft for now, I'm going to promote our current asan job to regular, and then once I've made that change, you can edit this PR so that the "just the part that builds the C runtime with tsan" job goes into the regular pipeline in the same file as the asan job..

File: .buildkite/pipelines/experimental/misc/sanitizers.yml
Comment (DilumAluthge) [inline .buildkite/pipelines/experimental/misc/sanitizers.yml:29]: I'd recommend doing the following:

```
- JuliaCI/julia#v1:
    # Drop default "registries" directory, so it is not persisted from execution to execution
    persist_depot_dirs: packages,artifacts,compiled
    version: '1.6'
```

File: contrib/tsan/Make.user.tsan
Comment (vchuravy) [reply contrib/tsan/Make.user.tsan:10]: We can also build LLVM with ccache on CI.

Comment (DilumAluthge) [reply contrib/tsan/Make.user.tsan:10]: > We can also build LLVM with ccache on CI.

When @staticfloat and I last discussed this, our intention was that the Base Julia CI should never be building dependencies from source.

Comment (DilumAluthge) [reply contrib/tsan/Make.user.tsan:10]: I think a `LLVM_tsan_jll` makes the most sense. Then it only needs to be built very occasionally in Ygg, instead of building LLVM from source on every PR that is opened to JuliaLang/julia.

Comment (vchuravy) [reply contrib/tsan/Make.user.tsan:10]: I don't really want to support "yet" another version of LLVM in Yggdrasil and slow the update cycle down even more.

Comment (DilumAluthge) [reply contrib/tsan/Make.user.tsan:10]: Hmmm. That's a fair point. Let's see if @staticfloat has any ideas.

At least we can keep the first part, which doesn't require building LLVM from source.

File: .buildkite/pipelines/experimental/misc/sanitizers.yml
Comment (tkf) [inline .buildkite/pipelines/experimental/misc/sanitizers.yml:31]: Ref:
* https://github.com/JuliaLang/julia/pull/42475

File: contrib/tsan/Make.user.tsan
Comment (tkf) [reply contrib/tsan/Make.user.tsan:10]: What's the source of slowness in yet another LLVM JLL vs ccache? Aren't they equivalent in terms of machine time? Is it that, with the JLL approach, there need to be humans involved in the process waiting for the JLL and updating the URL/version/checksum? If that's the case, it can also potentially be solved by more automation?

I kinda liked the JLL approach since it'd make it easier for more casual users to try sanitizers. OTOH, I guess the ccache approach would be more hackable for core devs since it'd give us a "healthy" CI'ed build script that is easy to flip some LLVM build options. Maybe the upside of the ccache approach wins, since there's no single set of build options that covers all the needs when you debugging things.

Comment (vchuravy) [reply contrib/tsan/Make.user.tsan:10]: > Is it that, with the JLL approach, there need to be humans involved in the process waiting for the JLL and updating the URL/version/checksum? If that's the case, it can also potentially be solved by more automation?

Yes, the maintenance of the LLVM JLL is a lot of human time, some of it waiting, some of it making sure that all the version number bits line up. Sheparding things through to the registry. 

I made attempts at trying to automate it, but currently it is pareto optimal in that I find is really annoying, but not annoying enough to descend to arcane wizardy. c.f https://github.com/JuliaPackaging/Yggdrasil/pull/726#issuecomment-617469147

**Comment by DilumAluthge** [comment]
For now, let's just do the first part, which requires no LLVM build. This will need a rebase on the latest master.

**Review by DilumAluthge — APPROVED** [review]

**Comment by DilumAluthge** [comment]
We should remember to squash-merge when merging this.

**Comment by DilumAluthge** [comment]
Assuming that the `tsan` job passes on Buildkite, this looks good to me to merge.

**Comment by DilumAluthge** [comment]
@tkf What Julia versions will tsan work on? Should we do the same thing we did for asan, and only run it on 1.8 and later? See the asan configuration for details, but basically you add a conditional of the form:
```yaml
    if: | # We only run the `asan` job on Julia 1.8 and later.
      (pipeline.slug != "julia-release-1-dot-6") && (pipeline.slug != "julia-release-1-dot-7")
```

**Comment by tkf** [comment]
Yeah, it's #42440 (so 1.8) and later.

(We might be able to find some ranges of commits that TSAN could be built but I think it'd require some serious archaeologies.)

**Comment by tkf** [comment]
I'm getting this error https://buildkite.com/julialang/julia-master/builds/4307#a956be25-a524-450d-b6f2-3c71b7cfb920/167-238

```
make[1]: Entering directory '/cache/build/amdci7-1/julialang/julia-master/src/support'
# potential confusion with debugging tools, when inspecting a process that has both `libjulia` and `libjulia-internal`
make -C /cache/build/amdci7-1/julialang/julia-master/src/support debug BUILDDIR='/cache/build/amdci7-1/julialang/julia-master/tmp/test-tsan/tsan/src/support'
make: *** No rule to make target '/cache/build/amdci7-1/julialang/julia-master/tmp/test-tsan/tsan/usr/lib/libuv.a', needed by '/cache/build/amdci7-1/julialang/julia-master/tmp/test-tsan/tsan/usr/lib/libjulia-internal-debug.so.1.8'.  Stop.
```

The error `No rule to make target '/.../usr/lib/libuv.a,` is a bit puzzling. I wounder if it's due to that I'm using `make -C src` on an out-of-source build?

**Comment by DilumAluthge** [comment]
Looks like you fixed it?

**Comment by tkf** [comment]
Yeah, it looks like `julia-src-debug` on the build root did the trick. (I'm guessing we need to pass around some states from the root `Makefile` to `src/Makefile`.)

**Comment by DilumAluthge** [comment]
> (I'm guessing we need to pass around some states from the root `Makefile` to `src/Makefile`.)

Do you want to make those fixes in this PR, or a separate PR?

**Comment by tkf** [comment]
Oh, I was guessing the reason why `make julia-src-debug` worked in the out-of-tree build while `make -C src debug` broke something. AFAICT `julia-src-debug` just wraps `make -C src debug`

https://github.com/JuliaLang/julia/blob/5a1b8cea6822c28a05dea4027585430a746f4998/Makefile#L75-L76

I think this PR is good to go, unless you want to double-check `julia-src-debug` is the correct make target.

**Comment by DilumAluthge** [comment]
> unless you want to double-check `julia-src-debug` is the correct make target

@Keno or @vchuravy Can you confirm that this is the correct target to be building?

Once they've confirmed that, let's merge this.

**Comment by vchuravy** [comment]
Since we are setting `override JULIA_BUILD_MODE=debug` in `Make.user` yes that should be right.

**Pull request merged**
